File: examples/webhook-ngrok/routes/proxy.py
import base64

import httpx
from fastapi import APIRouter, Request, Response

from auth import extract_auth_headers, verify_hmac, verify_ticket
from config import error_response, log
from crypto import decrypt_token_field, get_encryption_key
from middleware import safe_json_loads
from store import kv_store

router = APIRouter()


@router.post("/v1/proxy")
async def proxy(request: Request):
    """MCP Proxy endpoint — webhook injects credential into upstream request.

    TV forwards the agent's request here with a signed proxy ticket.
    The webhook decrypts the credential, injects it into the upstream
    request headers (replacing ${TOKEN}), makes the upstream call,
    and returns the response. TV never sees the credential.
    """
    rid, body_bytes, sig, ts, req_id = extract_auth_headers(request)
    log.info("proxy_enter rid=%s", rid)

    if body_bytes is None:
        body_bytes = await request.body()

    # Verify HMAC headers (authenticates TV)
    auth_err = verify_hmac(body_bytes, sig, ts, req_id, rid)
    if auth_err is not None:
        return auth_err

    data, err = safe_json_loads(body_bytes)
    if data is None:
        return error_response(400, "invalid_request", f"Invalid JSON: {err}")

    ticket = data.get("ticket", "")
    service = data.get("service", "")
    upstream = data.get("upstream", {})
    header_templates = data.get("headerTemplates", {})

    if not ticket or not service:
        return error_response(400, "invalid_request", "Missing 'ticket' or 'service'")
    if not upstream.get("url"):
        return error_response(400, "invalid_request", "Missing 'upstream.url'")

    # Verify ticket (authorizes the specific proxy operation)
    payload, ticket_err = verify_ticket(ticket, rid)
    if ticket_err is not None:
        return ticket_err

    if payload.get("pur") != "proxy":
        return error_response(401, "ticket_invalid", "Ticket purpose must be 'proxy'")
    if payload.get("svc") != service:
        return error_response(401, "ticket_invalid", f"Ticket is for service '{payload.get('svc')}', not '{service}'")

    # Decrypt the credential
    key = get_encryption_key()
    stored_doc = kv_store["tokens"].get(service)
    if not stored_doc:
        return error_response(404, "token_not_found", f"No token stored for service '{service}'")

    try:
        if "fields" in stored_doc:
            access_token = decrypt_token_field(key, stored_doc, "accessToken", rid)
        else:
            # Plaintext format (stored via backend /v1/storage)
            access_token = stored_doc.get("accessToken")
        if not access_token:
            return error_response(404, "token_not_found", f"No accessToken found for service '{service}'")
    except Exception as e:
        log.exception("proxy_decrypt_failed rid=%s", rid)
        return error_response(500, "internal_error", f"Credential decryption failed: {e}")

    # Build upstream request with credential injected
    upstream_url = upstream["url"]
    upstream_method = upstream.get("method", "GET").upper()
    upstream_headers = dict(upstream.get("headers", {}))
    upstream_body_b64 = upstream.get("body")
    upstream_body = base64.b64decode(upstream_body_b64) if upstream_body_b64 else None

    # Inject credential into header templates
    for header_name, template in header_templates.items():
        upstream_headers[header_name] = template.replace("${TOKEN}", access_token)

    log.info(
        "proxy_upstream rid=%s method=%s url=%s",
        rid, upstream_method, upstream_url,
    )

    # Make upstream request
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            upstream_resp = await client.request(
                method=upstream_method,
                url=upstream_url,
                headers=upstream_headers,
                content=upstream_body,
            )

        # Return upstream response
        resp_headers = {
            "X-Upstream-Status": str(upstream_resp.status_code),
        }

        log.info(
            "proxy_exit rid=%s upstream_status=%s",
            rid, upstream_resp.status_code,
        )

        return Response(
            content=upstream_resp.content,
            status_code=upstream_resp.status_code,
            headers=resp_headers,
            media_type=upstream_resp.headers.get("content-type", "application/json"),
        )

    except httpx.TimeoutException:
        return error_response(504, "upstream_timeout", "Upstream request timed out")
    except Exception as e:
        log.exception("proxy_upstream_failed rid=%s", rid)
        return error_response(502, "upstream_error", f"Upstream request failed: {e}")

File: examples/webhook/src/modules/proxy.ts
// MCP proxy (routes/proxy.py). TV forwards the agent's upstream request here
// with a signed proxy ticket; the webhook decrypts the credential, substitutes
// ${TOKEN} into the caller's header templates, calls the upstream verbatim, and
// streams the response straight back. TV never sees the credential. Authn is
// BOTH HMAC (TV identity) and the ticket (per-operation authorization).

import { hmacAuth } from "../core/middleware/hmacAuth.ts";
import { readJsonBody } from "../core/middleware/body.ts";
import { sendError } from "../core/middleware/respond.ts";
import { base64Decode } from "../core/crypto/encoding.ts";
import { invalidRequest, ticketInvalid, tokenNotFound, upstreamError, upstreamTimeout } from "../core/protocol/errors.ts";
import { verifyTicket } from "../core/protocol/tickets.ts";
import { decryptTokenField } from "../core/protocol/tokendoc.ts";
import type { FeatureModule } from "../core/registry.ts";
import { isGcpServiceAccount, mintGcpAccessToken } from "./interceptors/gcpSa.ts";

// Hop-by-hop headers not forwarded from the upstream response (proxy.py:118).
const HOP_BY_HOP = new Set(["content-encoding", "content-length", "transfer-encoding", "connection"]);
const UPSTREAM_TIMEOUT_MS = 30_000;

interface UpstreamSpec {
  url?: string;
  method?: string;
  headers?: Record<string, string>;
  body?: string | null;
}

export function proxyModule(): FeatureModule {
  return {
    name: "proxy",
    capability: "proxy",
    register(app, ctx) {
      app.post("/v1/proxy", hmacAuth(ctx), async (c) => {
        try {
          const body = readJsonBody(c);
          const ticket = typeof body.ticket === "string" ? body.ticket : "";
          const service = typeof body.service === "string" ? body.service : "";
          const upstream = (body.upstream && typeof body.upstream === "object" ? body.upstream : {}) as UpstreamSpec;
          const headerTemplates =
            body.headerTemplates && typeof body.headerTemplates === "object"
              ? (body.headerTemplates as Record<string, string>)
              : {};

          if (!ticket || !service) return sendError(c, invalidRequest("Missing 'ticket' or 'service'"));
          if (!upstream.url) return sendError(c, invalidRequest("Missing 'upstream.url'"));

          const secret = await ctx.secrets.hmacSecret();
          const payload = await verifyTicket(ticket, secret, ctx.replay);
          if (payload.pur !== "proxy") return sendError(c, ticketInvalid("Ticket purpose must be 'proxy'"));
          if (payload.svc !== service) {
            return sendError(c, ticketInvalid(`Ticket is for service '${payload.svc}', not '${service}'`));
          }

          const storedDoc = await ctx.storage.get("tokens", service);
          if (!storedDoc) return sendError(c, tokenNotFound(`No token stored for service '${service}'`));

          const key = await ctx.secrets.encryptionKey();
          let accessToken: string | null;
          if ("fields" in storedDoc) {
            accessToken = await decryptTokenField(key, storedDoc, "accessToken");
          } else {
            accessToken = storedDoc.accessToken != null ? String(storedDoc.accessToken) : null;
          }
          if (!accessToken) return sendError(c, tokenNotFound(`No accessToken found for service '${service}'`));

          // GCP SA key → mint a short-lived token and inject that instead.
          if (isGcpServiceAccount(accessToken)) {
            const minted = await mintGcpAccessToken(accessToken, service);
            accessToken = minted.accessToken;
          }

          const headers = new Headers();
          for (const [k, v] of Object.entries(upstream.headers ?? {})) headers.set(k, v);
          for (const [name, template] of Object.entries(headerTemplates)) {
            headers.set(name, template.replaceAll("${TOKEN}", accessToken));
          }

          const method = (upstream.method ?? "GET").toUpperCase();
          const reqBody = upstream.body ? base64Decode(upstream.body) : undefined;

          let upstreamResp: Response;
          try {
            upstreamResp = await fetch(upstream.url, {
              method,
              headers,
              ...(reqBody ? { body: reqBody as Uint8Array<ArrayBuffer> } : {}),
              signal: AbortSignal.timeout(UPSTREAM_TIMEOUT_MS),
            });
          } catch (err) {
            if (err instanceof DOMException && err.name === "TimeoutError") {
              return sendError(c, upstreamTimeout("Upstream request timed out"));
            }
            return sendError(c, upstreamError(`Upstream request failed: ${err instanceof Error ? err.message : String(err)}`));
          }

          const respHeaders: Record<string, string> = {};
          upstreamResp.headers.forEach((value, k) => {
            if (!HOP_BY_HOP.has(k.toLowerCase())) respHeaders[k] = value;
          });
          respHeaders["X-Upstream-Status"] = String(upstreamResp.status);

          const payloadBytes = new Uint8Array(await upstreamResp.arrayBuffer());
          return c.body(payloadBytes, upstreamResp.status as 200, respHeaders);
        } catch (e) {
          return sendError(c, e);
        }
      });
    },
  };
}
